### PR TITLE
Define gateway to node rpc in proto3 spec

### PR DESCRIPTION
**Pull request by okjodom**
#1103 proposes an architecture that decouples the gateway logic from the node used to process lightning payments. From that proposal, we need to define an RPC interface between the gateway and the node. This PR is an early definition for such an RPC interface using [gRPC](https://grpc.io/) protocol buffers ([proto3](https://grpc.io/))

### CODE REVIEW

**Comment by dpc** [comment]
Must set a env var to point at PROTOC

https://docs.rs/prost-build/latest/prost_build/#sourcing-protoc

Like here https://github.com/fedimint/fedimint/blob/16b5889fcc11716a3d9932ed8d58b91d51ff261d/flake.nix#L321 we do for other stuff.

**Comment by okjodom** [comment]
> Must set a env var to point at PROTOC
> 
> https://docs.rs/prost-build/latest/prost_build/#sourcing-protoc
> 
> Like here
> 
> https://github.com/fedimint/fedimint/blob/16b5889fcc11716a3d9932ed8d58b91d51ff261d/flake.nix#L321
> we do for other stuff.

This pointer helped get a working compiler to build! Applied fix at [commit 909dd455](https://github.com/fedimint/fedimint/pull/1143/commits/909dd4553fec8d7b49622d9d8fd76f602c3f808a)

Now seeing an issue where compile doesn't detect the relevant protobuf files on relative paths provided. I suspected it's because of how we filter files on the derivation, so my attempt to fix is on [commit #2550cf3d](https://github.com/fedimint/fedimint/pull/1143/commits/2550cf3dfb78276150392b93908f769b0147889b). It's a non fix.

@dpc do you have any pointer to this new issue? See https://github.com/fedimint/fedimint/actions/runs/3818394666/jobs/6495267730#step:5:3115

**Comment by dpc** [comment]
@okjodom There's one more place https://github.com/fedimint/fedimint/blob/master/flake.nix#L233

You can debug if it's working by uncommenting these two lines:

https://github.com/fedimint/fedimint/blob/16b5889fcc11716a3d9932ed8d58b91d51ff261d/flake.nix#L236
https://github.com/fedimint/fedimint/blob/16b5889fcc11716a3d9932ed8d58b91d51ff261d/flake.nix#L277

**Comment by dpc** [comment]
BTW. Nowhere else in fedimint we're using protobuf, is there any reason to use them now? Seems like we wanted to keep it lean and use encodable traits as much as we can, and only when we can not, revert to bincode.

A crate with encodeable types should probably be our API definition.

**Comment by okjodom** [comment]
> Nowhere else in fedimint we're using protobuf, is there any reason to use them now?

This was my reasoning for adopting protobufs for the rpc definition: 
image
ref: https://discord.com/channels/990354215060795454/999362050625130656/1053791066098245663

> Seems like we wanted to keep it lean and use encodable traits as much as we can, and only when we can not, revert to bincode. **A crate with encodeable types should probably be our API definition.**

I think we can compare the merits or limitations of both patterns here i.e I haven't tested the encodable trait library approach, though I can clearly see the benefits offered  by protobufs. Marking the conversation as unresolved so we can revisit, but at this stage I'd continue with implementation based on protobufs so we can get a gateway version out

File: gateway/ln-gateway/proto/gatewaylnrpc.proto [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  // Subscribe to intercepted HTLCs with this short channel id
+  uint64 short_channel_id = 1;
+}
+
+message SubscribeInterceptHtlcsResponse {
+  // The payment hash of the invoice
+  bytes payment_hash = 1;
+
+  // The amount of the invoice
+  uint64 amount = 3;
```
Comment (m1sterc001guy) [inline gateway/ln-gateway/proto/gatewaylnrpc.proto:51]: Intentionally skipping 2?

Comment (okjodom) [reply gateway/ln-gateway/proto/gatewaylnrpc.proto:51]: no. did not mean to reserve that slot. Fix incoming

**Comment by elsirion** [comment]
This proliferation of different RPC protocol isn't great, but we'll need gRPC for LND anyway, so maybe not the biggest problem.

File: gateway/ln-gateway/proto/gatewaylnrpc.proto [hunk 2]
```diff
@@ -0,0 +1,58 @@
+syntax = "proto3";
```
Comment (elsirion) [inline gateway/ln-gateway/proto/gatewaylnrpc.proto:1]: I wonder if we can just emulate LND with the CLN plugin such that we just need one impl in the gateway daemon?

**Comment by elsirion** [comment]
Not sure what the outcome of the call was, don't want to needlessly block but we do have too many RPC protocols … @dpc @justinmoon @jkitman I'm ok with merging.

**Review by elsirion — APPROVED** [review]
Chatted with @okjodom, moving forward with protobuf for now

**Pull request merged**